core/corehttp/gateway_handler.go

```diff
@@ -163,23 +194,170 @@ func (i *gatewayHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 func (i *gatewayHandler) postHandler(w http.ResponseWriter, r *http.Request) {
 	nd, err := i.NewDagFromReader(r.Body)
 	if err != nil {
-		w.WriteHeader(http.StatusInternalServerError)
-		log.Error(err)
-		w.Write([]byte(err.Error()))
+		internalWebError(w, err)
 		return
 	}
 
 	k, err := i.AddNodeToDAG(nd)
 	if err != nil {
-		w.WriteHeader(http.StatusInternalServerError)
-		log.Error(err)
+		internalWebError(w, err)
+		return
+	}
+
+	http.Redirect(w, r, "/ipfs/"+mh.Multihash(k).B58String(), http.StatusCreated)
+}
+
+func (i *gatewayHandler) putEmptyDirHandler(w http.ResponseWriter, r *http.Request, path string) {
+	newnode := NewDagEmptyDir()
+
+	key, err := i.node.DAG.Add(newnode)
+	if err != nil {
+		webError(w, "Could not add recursively new node", err, http.StatusInternalServerError)
+		return
+	}
+
+	http.Redirect(w, r, "/ipfs/"+key.String()+"/", http.StatusCreated)
+}
+
+func (i *gatewayHandler) putHandler(w http.ResponseWriter, r *http.Request, path string) {
```

Comment:
this function is pretty long and there's lots of edge-case handling. Let's add some comments like so:

```go
// resolving path components into merkledag nodes
// if no links are found, create empty directories -- which will be linked and populated below.
path_nodes, err := i.node.Resolver.ResolveLinks(rootnd, components[:len(components)-1])
if _, ok := err.(p.ErrNoLink); ok {
  // Create empty directories, links will be made further down the code
  for len(path_nodes) < len(components) {
    path_nodes = append(path_nodes, NewDagEmptyDir())
  }
} else if err != nil {
  webError(w, "Could not resolve parent object", err, http.StatusBadRequest)
  return
}
```